### PR TITLE
Plane: added blackbox auto-arming feature

### PR DESCRIPTION
**Pull request by tridge**
This is a feature for data loggers for F3A planes like this one:
https://www.ebay.com/itm/165147365288
it is only selectable in the custom build server. When enabled plane will auto-arm when the speed goes above the parameter threshold ARMING_BBOX_SPD, and disarm when below that speed for 20 seconds.
The reason for not just doing LOG_DISARMED is we need the EKF to think we are armed, to get accurate logging of state information

### CODE REVIEW

**Comment by peterbarker** [comment]
This *really* can't be scripting?!

**Review by peterbarker — APPROVED** [review]
LGTM

File: ArduPlane/AP_Arming.cpp [hunk 1]
```diff
@@ -387,6 +398,26 @@ void AP_Arming_Plane::update_soft_armed()
...
+    if (blackbox_speed > 0) {
+        const float speed3d = plane.gps.status() >= AP_GPS::GPS_OK_FIX_3D?plane.gps.velocity().length():0;
+        const uint32_t now = AP_HAL::millis();
+        if (speed3d > blackbox_speed) {
+            last_over_3dspeed_ms = now;
+        }
+        if (!_armed && speed3d > blackbox_speed) {
+            // force safety on so we don't run motors
+            hal.rcout->force_safety_on();
+            arm(Method::BLACKBOX, false);
```
Comment (IamPete1) [inline ArduPlane/AP_Arming.cpp:412]: I suspect this doesn't work anymore now we have a mandatory RC check, or at least there is some more setup required.

**Comment by tridge** [comment]
> This _really_ can't be scripting?!

these boards are mostly F405, can't do scripting

File: ArduPlane/AP_Arming.cpp (hunk 1 above)
Comment (tridge) [reply ArduPlane/AP_Arming.cpp:412]: good point, I'll test in SITL and add an autotest to make sure it keeps working

**Comment by peterbarker** [comment]
> > This _really_ can't be scripting?!
> 
> these boards are mostly F405, can't do scripting

I *did* that test.  f405 can run scripting - at least for trivial scripts like this would be.

**Comment by tridge** [comment]
> I _did_ that test. f405 can run scripting - at least for trivial scripts like this would be.

with EKF3, GPS drivers etc and enough log buffer space to guarantee we get the log? The point of the blackbox logger is to run EKF3 with logging

**Comment by peterbarker** [comment]
This commit shows LUA running on an omnibusf4 (F405).

There's 5kB of free RAM while running the script.

Scripting is started with a 50kB heap size.

https://github.com/ArduPilot/ardupilot/compare/master...peterbarker:ardupilot:wip/f405-plane-scripting

The script itself is pretty-much untested - it's as far as `AP: Lua: @ROMFS/scripts/blackbox-logging.lua:19: attempt to compare Parameter with nil` - I don't have a GPS attached :-)

I haven't even really done huge amounts of work here to scrounge more RAM - I'm sure more can be found!

That branch has to be compiled with `--enable-scripting` on the configure command-line.

**Comment by tridge** [comment]
> This commit shows LUA running on an omnibusf4 (F405).

we need both EKF2 and EKF3 running (yes, really is needed, I can send a log to show why if interested)

**Comment by IamPete1** [comment]
Why would we need EKF2 and EKF3 here but not on "real" vehicles?

If we can do it in scripting I would much prefer that. Would need a little more work on the build server, but we were talking about that anyway, so it could enable scripting and put the script in ROMFS as one button.

**Comment by tridge** [comment]
> Why would we need EKF2 and EKF3 here but not on "real" vehicles?

because aerobatics, without airspeed, with bad compass, no arming checks and no GCS often leads to very poor state estimates. Having EKF2 and EKF3 greatly increases the chance of a useful log
we're servicing a very different community with this feature

**Pull request merged**